### ISSUE TITLE
Command line needs few enhancements

### ISSUE THREAD

**Issue by AstroDeepSky** [Steffenhir/GraXpert#95]
Hi,

I'm using command line mode to run GraXpert from PixInsight with a dedicated small javascript.
Script is very simple:
1. Save target view in Pix temporary directory (XISF)
2. Launch GraXpert
3. Open result file (XISF) in Pix

I have 2 main issues... 
1. GraXpert open UI (welcome GraXpert... Release Ariel...) delaying the image processing (It takes more time for GraXpert to open UI versus image processing).
2. GraXpert is launched in background which is not ideal for scripting (need to implement a loop after call to GraXpert, no possibility to get exit code in case or error, ...).

Let me know if you need more details.

I'm currently using Python version to avoid both issues.
I'll report other problems related to Python.

**Comment by killerciao** [comment]
Can you share the script ? I'd love to have an implementation inside Pixinsight of GraXpert but i don't know anything about coding :D

**Comment by Joel95** [comment]
Sure, give me few days to make some clean-up and write a small readme file
to explain how to install it inside PixInsight and make the connection to
GraXpert.
Script also restore metadata to fix
https://github.com/Steffenhir/GraXpert/issues/70 and
https://github.com/Steffenhir/GraXpert/issues/85
Attached a screenshot.

On your side, could you have a look on
https://github.com/Steffenhir/GraXpert/issues/95 ?
It will improve a lot the integration inside PixInsight.

Regards,

Joël
.

**Comment by Joel95** [comment]
Hello,

Attached zip file with script and ReadMe inside with all details.
As you'll see my script make a call to GraXpert and display result inside
PixInsight.
I have implemented all command line options available in last GraXpert
version 2.0.2.
If you could put few changes I did in official release of GraXpert (all is
described in ReadMe).

Regards,

Joël
 PixGraXpert-v0.01.zip
<https://drive.google.com/file/d/1aysHcC06q4YBUv5JGOd8NgkH7yYgF0w2/view?usp=drive_web>

**Comment by Joel95** [comment]
Hi,

Here is a new package for PixInsight on Windows:
- Full command line interface of GraXpert for better integration in PixInsight
- PixGraXpert, a PixInsight script featuring GraXpert

PixGraXpert includes
- Drop down list for the selection of the image to process
- Drop down list for the selection of AI model (1.0.1 and 1.0.0 tested)
- Drop down list for the selection of correction (Subtraction or Division)
- Full control of Smoothing (0 to 1)
- Replace target image or create a new picture
- Restore all metadata from original image (**astrometric data preserved**)
- Create and open background model in PixInsight
- Debug option (PixGraXpert and GraXpert logs in PixInsight console window)
- Save PixGraXpert script as standard process icon on PixInsight workspace
- Execute PixGraXpert script as standard process icon from PixInsight workspace

More details with screenshots here:
https://drive.google.com/file/d/1_EMgjJcO7r85GRb9Mua2UDCjiCnPD0gj/view?usp=sharing

Package including source files available here:
https://drive.google.com/file/d/1TgYec37W2XJx7JogrKo67jX251KkoauF/view?usp=sharing

Enjoy ;-)

**Comment by killerciao** [comment]
Awesome thank you. You should submit it to Pixinsight forum ^_^

**Comment by Joel95** [comment]
@killerciao this is done :-)
https://pixinsight.com/forum/index.php?threads/pixgraxpert-run-graxpert-from-pix-ai-gradient-extraction-removal.22147/
Let see feedbacks...

**Comment by killerciao** [comment]
Pixinsight forum... damn they are stricty. They removed your post. Nonsense.

> @killerciao this is done :-) https://pixinsight.com/forum/index.php?threads/pixgraxpert-run-graxpert-from-pix-ai-gradient-extraction-removal.22147/ Let see feedbacks...

**Comment by schmelly** [comment]
Hi @Joel95,

thanks for your contribution and ideas and sorry for the delayed answer: GraXpert is a hobby for us and at the moment we are busy in our jobs. For a more interactive conversation I suggest you join our Discord "Dark Matters" here (if not done already): https://discord.com/channels/765322305546747935/961278926909497424

Four your information:
- I am currently in the process of migrating our packaging from PyInstaller to cx_freeze. One of the main reasons is that the startup time has increased significantly since we included tensorflow as a dependency: anti malware engines will scan the application each time after startup. The effect of this transition will be twofold: 1) the splash screen won't be supported anymore and 2) GraXpert will be packaged as .msi on Windows, .dmg on macOS and .rpm+.deb on Linux. This should address your main issues with the cli interface. Of course further contributions are welcome but we are still a bit hesitant to provide different distributions for gui- and cli- based version of GraXpert to not confuse our users (many have no strong IT background).
- There is another script-based integration for GraXpert AI in PixInsight already as part of the PixInsight Toolbox of Juergen Terpe: https://pixinsight.com/forum/index.php?threads/new-scripts-pixinsight-toolbox.19401/
I by myself do not own PixInsight. Hence, I cannot participate in this development much. However, please mind that Juergen is also a member of our "Dark Matters" Discord community and we are in frequent contact with him. We probably won't be able to actively support different script integrations in the long run. Hence, I highly suggest you get in contact with Juergen to align development efforts.

Another plea from my side: I was able to read your posts in the PixInsight forum prior to deletion. If you adapt GraXpert and publish it by yourself that is perfectly fine - in the end it is Open Source. But please make sure it is recognizable as derivative work as I don't want to do support for derivatives on top of the a-bit-slow-one we have in place for our releaeses ;-)

Best regards and CS
David

**Comment by Joel95** [comment]
Thank you, David, for your response and the valuable insights you provided.
Development is also a hobby for me.

I've joined the Discord, but a significant portion of it is in German.
I hope the migration towards cx_freeze will address the major issue.
I'll stay tuned.

Yes, I don't want to bring confusion with a derivative work.
In reality, I've made very few minor changes that, in a way, don't disrupt
the existing implementation.
Some of these changes are specific to the command line.

Here's the list of changes:
1. Automatic download of AI models disabled if graxpert.s3_secrets is not
provided.
2. update_xisf_imagedata returns in case no fits_header is found (useful
for saving the background model, for instance).
3. New command line parameter "-output" for specifying the output filename.
4. Output XISF file when the input file is XISF (Command Line Interface
only).
5. Output of the background model (Command Line Interface only).
6. Changes in logs format for better integration into PixInsight console
(Command Line Interface only).
7. Reformatted standard error for better display (buffer output until the
end of lines).

I hope you'll consider incorporating these in a future version ;-)

I'm already in touch with Juergen Terpe and would be happy to contribute if
he needs assistance.

Best regards,

Joël
